### PR TITLE
[JENKINS-30600] demonstrate fix in git-client

### PR DESCRIPTION
**Pull request by ndeloof**

### CODE REVIEW

**Comment by ndeloof** [comment]
this PR introduce a test case to demonstrate https://github.com/jenkinsci/git-client-plugin/pull/186 is fixing Launcher issue. @reviewbybees

**Comment by jglick** [comment]
You forgot to `mvn -DskipTests deploy` the upstream PR.

File: src/test/java/hudson/plugins/git/GitSCMTest.java [hunk 1]
```diff
@@ -2113,7 +2115,33 @@ public void testBuildEnvVarsLocalBranchNotSet() throws Exception {
...
-    
+
+    @TestExtension("testClientUsesLauncher")
+    public static final LauncherDecorator DECORATOR = new LauncherDecorator() {
+        @Override
+        public Launcher decorate(Launcher launcher, Node node) {
+            return new Launcher.DecoratedLauncher(launcher) {
+                @Override
+                public Proc launch(ProcStarter starter) throws IOException {
+                    starter.cmds("wil fail");
```
Comment (jglick) [inline src/test/java/hudson/plugins/git/GitSCMTest.java]: :ant: typo

File: src/test/java/hudson/plugins/git/GitSCMTest.java [hunk 2]
```diff
@@ -2113,7 +2115,33 @@ public void testBuildEnvVarsLocalBranchNotSet() throws Exception {
...
+
+    @Issue("JENKINS-30600")
+    @Test
+    public void testClientUsesLauncher() throws Exception {
+        final FreeStyleProject project = setupProject("*/*", false);
+        final String commitFile1 = "commitFile1";
+        commit(commitFile1, johnDoe, "Commit number 1");
+
+        final FreeStyleBuild build = build(project, Result.FAILURE);
+        assertTrue(build.getLog().contains("Cannot run program \"wil fail\""));
```
Comment (jglick) [inline src/test/java/hudson/plugins/git/GitSCMTest.java]: :ant: to match

Comment (jglick) [inline src/test/java/hudson/plugins/git/GitSCMTest.java]: anyway you want

```java
rule.assertLogContains("will fail", build);
```

**Comment by jglick** [comment]
:bee:

**Comment by ndeloof** [comment]
> You forgot to mvn -DskipTests deploy the upstream PR.
This makes no sense, doing that I would deploy some SNAPSHOT that could be overridden at any time by anyone doing another deploy. Until we have some sane way to test cross repository PR I don't see this as a reasonable way to valide.
Anyway, I've discovered a regression on jenkinsci/git-client-plugin#186 that makes it a no-go, and can't find how this could be fixed.

**Comment by jglick** [comment]
> doing that I would deploy some SNAPSHOT that could be overridden at any time by anyone doing another deploy

In practice that is uncommon, but if you are concerned about the possibility, you need merely refer to the timestamp in the downstream PR’s `<version>`. This ensures that the downstream PR is always building against exactly the version you deployed. The downside is that if you make significant changes to the upstream PR you must remember to not just re`deploy` them, but adjust the downstream PR to refer to the new timestamps.

**Comment by ndeloof** [comment]
obsolete

**Pull request closed without merging**